trackpy/static.py

```diff
@@ -0,0 +1,264 @@
+    cutoff : float
+        Maximum distance to calculate g(r)
+    fraction : float, optional
+        The fraction of particles to calculate g(r) with. May be used to increase speed of function.
+        Particles selected at random.
+    dr : float, optional
+        The bin width
+    p_indices : sequence, optional
+        Only consider a pair of particles if one of them is in 'p_indices'.
+        Uses zero-based indexing, regardless of how 'feat' is indexed.
+    ndensity : float, optional
+        Density of particle packing. If not specified, density will be calculated assuming rectangular homogeneous
+        arrangement.
+    boundary : tuple, optional
+        Tuple specifying rectangular boundary of particles (xmin, xmax, ymin, ymax). Must be floats.
+        Default is to assume a rectangular packing. Boundaries are determined by edge particles.
+    handle_edge : boolean, optional
+        If true, compensate for reduced area around particles near the edges.
+
+    Returns
+    -------
+    r_edges : array
+        Return the bin edges
+    g_r : array
+        The values of g_r
+    """
+
+    if boundary is None:
+        xmin, xmax, ymin, ymax =  feat.x.min(), feat.x.max(), feat.y.min(), feat.y.max()
+    else:
+        xmin, xmax, ymin, ymax = boundary
+        # Disregard all particles outside the bounding box
+        feat = feat[(feat.x >= xmin) & (feat.x <= xmax) & (feat.y >= ymin) & (feat.y <= ymax)]
+
+    if ndensity is None:
+        ndensity = feat.x.count() / ((xmax - xmin) * (ymax - ymin))  #  particle packing density
+
+    if p_indices is None:
+        p_indices = random.sample(range(len(feat)), int(fraction * len(feat)))  # grab random sample of particles
+
+    r_edges = np.arange(0, cutoff + dr, dr)  # radii bins to search for particles
+    g_r = np.zeros(len(r_edges) - 1) 
+    max_p_count =  int(np.pi * (r_edges.max() + dr)**2 * ndensity * 10)  # upper bound for neighborhood particle count
+    ckdtree = cKDTree(feat[['x', 'y']])  # initialize kdtree for fast neighbor search
+    points = feat.as_matrix(['x', 'y'])  # Convert pandas dataframe to numpy array for faster indexing
+        
+    # For edge handling, two techniques are used. If a particle is near only one edge, the fractional area of the
+    # search ring r+dr is caluclated analytically via 1 - arccos(d / r ) / pi, where d is the distance to the wall. If
+    # the particle is near two or more walls, a ring of points is generated around the particle, and a mask is
+    # applied to find the the number of points within the boundary, giving an estimate of the area. Below,
+    # rings of size r + dr  for all r in r_edges are generated and cached for later use to speed up computation
+    n = 1000  # TODO: Should scale with radius, dr
+    refx, refy = _points_ring2D(r_edges, dr, n)
+
+    for idx in p_indices:
+        dist, idxs = ckdtree.query(points[idx], k=max_p_count, distance_upper_bound=cutoff)
```

Comment:
This would return a result set that scales with N^2, which for large systems can exceed available memory. So it might give a performance boost, but it should not be the default.